*: Use CPut when writing to `system.descriptor` table

Xiang-Gu opened the pull request:
Each individual commit has a rather thorough message, so I encourage
reviewers to read them as well.

Commit 1: code pattern, non-functional changes
Commit 2: Added a `[]byte` field in descriptor builder
and immutable struct. We also modified existing functions
to carry over this field between descriptor and Mutable/immutable
descriptor, and vice versa.

Commit 3: Upgrade `WriteDescToBatch` to use `CPut` rather than
`Put` to prevent clobbering of the `system.descriptor` table.
We get the expected value for the `CPut` when first reading the
descriptor into collection from storage, as well as keep booking
the descriptor's latest change inside the `uncommitted` descriptor
set, in case more we write to the same descriptor more than once
in the same transaction.

I didn't add any tests in this PR since this changed function 
`WriteDescToBatch` is heavily invoked for many sql stmts,
so I think we should be pretty confident about its correctness
if it passes all existing tests (unit, logic, and roachtest).

Release note: None

Xiang-Gu committed:
catalog: added two methods in interface Descriptor and DescriptorBuilder

This PR added a `[]byte` filed (called `rawBytesInStorage`) in the
[table|schema|database|type|function] immutable descriptor struct and
the [table|schema|database|type|function] descriptor builder struct.

The purpose is to faciliate the next commit where we want to do book
keeping of what's actually in storage of each descriptor, so that we can
use a CPut when updating a descriptor in storage. See next commit for
more details.

Accompanying this added field is two methods added:
 - `GetRawBytesInStorage() []byte` in `catalog.Descriptor`, and
 - `SetRawBytesInStorage([]byte)` in `catalog.DescriptorBuilder`
They serve as the getter and setter of this field.

We also modified method that gives a builder from a descriptor to carry
over this field.

The way this byte slice field will be initialized and carried over in
a builder and a descriptor will be something like the following graph,
where `XXX` denotes an entity, `xxx` denotes an action, and `==>`
denotes the direction of action.

      DESCPB.DESCRIPTOR
            | |
            | |
     construct a builder
            | |
            | |
             V
       CATALOG.BUILDER
            | |
            | |
      set byte slice field
            | |
            | |
             V
       CATALOG.BUILDER (with the byte slice field set)
            | |
            | |
        build a desc
            | |
            | |
             V
  MUTABLEDESCRIPTOR or DESCRIPTOR
  (with the byte slice field carried over from the builder)
            | |
            | |
         new builder
            | |
            | |
             V
      CATALOG.BUILDER
  (with the byte slice field carried over from the descriptor)
            | |
            | |
            ...

Xiang-Gu committed:
descs: upgrade `WriteDescToBatch` to use CPut

Previously, `WriteDescToBatch`, which is called to
update a descriptor in storage by a `desc.Collection`, uses `Put`
primitive to directly modify the storage layer. We would like to tighten
it to use a `CPut` to prevent (in)advertent clobbering of that table.

This PR does that by book-keeping the raw bytes of the to-be-updated
descriptor in the descriptor, acquired when we first read it into the
`desc.Collection` from the storage layer. Then, the infrastructural
work done in the previous commit allows us to carry over these raw bytes
as we are preparing the `MutableDescriptor` that we pass into this
`WriteDescToBatch` method.

One additional difficulty is that what if we will need to call
`WriteDescToBatch` more than once in one transaction. For example,
in the new schema changer, statement phase and precommit phase will both
be in one transaction, but we call `WriteDescToBatch` at the end of
each stage. Hence, for some DDL stmts (e.g. `DROP COLUMN`), we will
call `WriteDescToBatch` twice in one transaction. The first call will
modify the descriptor in storage and also added this descriptor to
`desc.Collection.uncommitted` set, so, the second call will get it from
there. To make `CPut` work correctly for the second call, we will need
to get the expected byte slice from the `uncommitted` descriptor set.
This motivates the change to update a descriptor's byte slice field when
it's added to the `uncommitted` descriptor sett.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/88844)
<!-- Reviewable:end -->

postamar commented on pkg/sql/catalog/descriptor.go:
```diff
@@ -88,6 +88,8 @@ type DescriptorBuilder interface {
 	// upgrade migrations
 	RunRestoreChanges(descLookupFn func(id descpb.ID) Descriptor) error
 
+	SetRawBytesInDescriptor(rawBytes []byte)
```
Can we pass the raw bytes to the `DescriptorBuilder` constructors instead? Either you create the builder from another descriptor and that byte slice is available there, or you create it from a protobuf and the bytes are usually not far. In any case, we have to pass them along.

I think this PR might benefit from being rebased on https://github.com/cockroachdb/cockroach/pull/88786 which tightens how these builders can be created.

Xiang-Gu replied:
The difficulty would be when we upsert a descriptor to the `uncommitted` descriptor set, we will need to update this `rawBytesInStorage` field, instead of passing it along. This is the main reason I choose to have a setter on this field.

postamar commented on pkg/sql/catalog/descs/uncommitted_descriptors.go:
```diff
@@ -186,7 +187,16 @@ func (ud *uncommittedDescriptors) upsert(
+	// in the future.
+	var val roachpb.Value
+	val.SetProto(mut.DescriptorProto())
+	rawBytesInStorage := val.TagAndDataBytes()
```
I'm sorry to say that this won't work. Protobuf serialization and deserialization is not bijective so you can't infer what the bytes in storage look like. 

A concrete example is when we remove certain fields in the descriptor proto definition: roundtripping like this will lose the bytes corresponding to the removed fields. For this to happen in a cluster you have to have a descriptor that hasn't changed over two major versions; which isn't as rare as you'd think.

postamar replied:
Turns out I got it wrong a bit, we discussed this offline.

Xiang-Gu committed:
funcdesc: changed function receiver to be pointer receiver

This will be needed when I later added a field to the function
builder which we wish to be able to set it.

Xiang-Gu commented on pkg/sql/catalog/tabledesc/structured.go:
```diff
@@ -49,7 +49,7 @@ import (
 // going through schema mutations.
 type Mutable struct {
-	wrapper
+	immutable
```
I agree! For now, I've reverted this commit. 

While I'm on this, I think I can have a PR that clears this part of code a bit, maybe after your https://github.com/cockroachdb/cockroach/pull/88786 is merged where we use the following pattern for all types of descriptors
```go
type wrapper struct {
    descpb.(Table|Schema|Database|Type|Function)Descriptor
    other_fields
}

type immutable struct {
   wrapper
   // no other fields should be here!
}

type Mutable struct {
  wrapper
  original *immutable
}
````
And, `wrapper` will be the base implementation for `catalog.Descriptor`, which means we should only have functions whose receivers are `*wrapper` and `*Mutable` but not `*immutable`.

Xiang-Gu commented:
@postamar I rebased and polished a bit more on your recent merge. I fixed a few test failures discovered from CI -- they're because I forgot to properly initialize this field for `unsafe_upsert_descriptor`.

 I think this is RFAL :)

postamar commented on pkg/sql/catalog/descs/collection.go:
```diff
@@ -269,6 +269,20 @@ func (tc *Collection) WriteDescToBatch(
+	// which means it has existed in `tc.uncommitted`, we will retrieve the expected
+	// bytes from there.
+	var expected []byte
+	if exist := tc.uncommitted.uncommitted.GetByID(desc.GetID()); exist == nil {
```
nit: use `getUncommittedByID` here instead

uber nit: perhaps it's just me but evaluating the `nil` case first in this if-block feels weird to me.

postamar commented on pkg/sql/catalog/descs/collection_test.go:
```diff
@@ -284,7 +284,7 @@ func TestAddUncommittedDescriptorAndMutableResolution(t *testing.T) {
 			require.NoError(t, err)
 			require.Equal(t, db.GetVersion(), immByNameAfter.GetVersion())
-			require.Equal(t, mut.ImmutableCopy(), immByNameAfter)
+			require.Equal(t, mut.ImmutableCopy().(catalog.DatabaseDescriptor).DatabaseDesc(), immByNameAfter.DatabaseDesc())
```
nit: if you want to avoid the cast here, you can also compare the results of `.DescriptorProto()` to the same effect.

postamar commented on pkg/sql/catalog/descs/uncommitted_descriptors.go:
```diff
@@ -186,7 +187,20 @@ func (ud *uncommittedDescriptors) upsert(
+	if err = val.SetProto(mut.DescriptorProto()); err != nil {
+		return err
+	}
+	rawBytesInStorage := val.TagAndDataBytes()
```
nit: rename this to `rawBytesInStorageAfterPendingWrite` perhaps?

postamar commented on pkg/sql/catalog/descs/uncommitted_descriptors.go:
```diff
@@ -186,7 +187,20 @@ func (ud *uncommittedDescriptors) upsert(
+	// compute what the raw bytes of this descriptor in storage is going to
+	// look like when that write happens, so that any further update to this
+	// descriptor, which will be retrieved from the "uncommitted" set, will
+	// carry the correct raw bytes in storage with it.
```
I'm fine with this, it's rather elegant.

postamar reviewed (approved):
LGTM conditional on addressing the minor comments. Nicely done.

Xiang-Gu reviewed (commented):
Thanks for the review! I've addressed the feedback.

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88844#-:-NDU3wSc1n9PDUBe47W5:bp66fhe)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar)

___
*[`pkg/sql/catalog/dbdesc/database_desc.go` line 352 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/88844#-ND-ExbHfrbCRWGl-Pud-r6-352:-NDTr82y0i9n54r-Ic3H:bvdg9ka) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f1bea1ef6e6c80c9be68ed0f21e969f60f6561e/pkg/sql/catalog/dbdesc/database_desc.go#L352)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Consider deep-copying the byte slice, either here or in the builder. Notice that the builder already deep-copies the protobuf in almost all cases except for an edge case on the hot path. The memory overhead is not a concern outside of this edge case.
</blockquote></details>

done. 
I've made it such that we deep copy this field when we
 - construct the builder from protobuf,
 - build a descriptor from the builder, and
- construct a new builder from a descriptor
___
*[`pkg/sql/catalog/descs/collection.go` line 280 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/88844#-ND-ExbIg7DAe2G9eYSH-r6-280:-NDTrvCA2yn_NSJgbdAi:blt4qd4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f1bea1ef6e6c80c9be68ed0f21e969f60f6561e/pkg/sql/catalog/descs/collection.go#L280)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: use `getUncommittedByID` here instead

uber nit: perhaps it's just me but evaluating the `nil` case first in this if-block feels weird to me.
</blockquote></details>

done for both nits!



_Code quote:_
```Go
if exist := tc.uncommitted.uncommitted.GetByID(desc.GetID())
```
___
*[`pkg/sql/catalog/descs/uncommitted_descriptors.go` line 196 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88844#-ND-ExbIg7DAe2G9eYSI-r3-196:-NDTs-Pp3dHCxrxSzrEB:bnlamys) ([raw file](https://github.com/cockroachdb/cockroach/blob/37b7e47f2b800e2615a754aa4583466aa6ce9480/pkg/sql/catalog/descs/uncommitted_descriptors.go#L196)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Turns out I got it wrong a bit, we discussed this offline.
</blockquote></details>

resolved!
___
*[`pkg/sql/catalog/descs/uncommitted_descriptors.go` line 195 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/88844#-ND-ExbIg7DAe2G9eYSI-r6-195:-NDTshYk20ew8-10Q5Bl:b2nid0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f1bea1ef6e6c80c9be68ed0f21e969f60f6561e/pkg/sql/catalog/descs/uncommitted_descriptors.go#L195)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I'm fine with this, it's rather elegant.
</blockquote></details>

thx
___
*[`pkg/sql/catalog/descs/uncommitted_descriptors.go` line 200 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/88844#-ND-ExbIg7DAe2G9eYSI-r6-200:-NDTtA0rA20bFFcxb-Yi:b-9r5ya) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f1bea1ef6e6c80c9be68ed0f21e969f60f6561e/pkg/sql/catalog/descs/uncommitted_descriptors.go#L200)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: rename this to `rawBytesInStorageAfterPendingWrite` perhaps?
</blockquote></details>

good suggestion! Done!
___
*[`pkg/sql/catalog/tabledesc/structured.go` line 1018 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/88844#-ND-ExbHfrbCRWGl-Pu_-r6-1018:-NDTtOFz-tPOKL5N8pty:bran39d) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f1bea1ef6e6c80c9be68ed0f21e969f60f6561e/pkg/sql/catalog/tabledesc/structured.go#L1018)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

`desc` can never be nil here, or in other `Descriptor` implementationgs
</blockquote></details>

I guess I was just paranoid when writing this -- I've removed the `if` block.



_Code quote:_
```Go
(desc *Mutable) GetRawBytesInStorage() []byte {
```
___
*[`pkg/sql/catalog/descs/collection_test.go` line 287 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/88844#-ND9CZB8X33QShYNfhMt-r6-287:-NDTu-Ea9aHUcX4jrKOO:bgup1mt) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f1bea1ef6e6c80c9be68ed0f21e969f60f6561e/pkg/sql/catalog/descs/collection_test.go#L287)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: if you want to avoid the cast here, you can also compare the results of `.DescriptorProto()` to the same effect.
</blockquote></details>

done, and I've also changed two other places where I did this.



_Code quote:_
```Go
require.Equal(t, mut.ImmutableCopy().(catalog.DatabaseDescriptor).DatabaseDesc()
```


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Nice work! My nits can be ignored or done later if you agree with them 

> Reviewed 1 of 21 files at r4, 1 of 8 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88844#-:-NDUBaJe7SO5yh0PxK5H:b-aq2p66)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @postamar, and @Xiang-Gu)

___
*[`pkg/sql/catalog/dbdesc/database_desc_builder.go` line 151 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/88844#-NDUB5Ib06k7j7MHtydl:-NDUB5Ib06k7j7MHtydm:b-flojt0) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa3c59ccb2836ebd04ae89a4a26d14d7ca1bbe0d/pkg/sql/catalog/dbdesc/database_desc_builder.go#L151)):*
> ```Go
> // SetRawBytesInStorage implements the catalog.DescriptorBuilder interface.
> func (ddb *databaseDescriptorBuilder) SetRawBytesInStorage(rawBytes []byte) {
> 	ddb.rawBytesInStorage = append([]byte(nil), rawBytes...) // deep-copy
> ```

nit: you can do `make([]byte, 0, len(rawBytes))` to allocate the right size array underneath here and elsewhere.
___
*[`pkg/sql/catalog/descs/uncommitted_descriptors.go` line 201 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/88844#-NDUBN9bByIQGHzyaw_W:-NDUBN9bByIQGHzyaw_X:brsyqed) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa3c59ccb2836ebd04ae89a4a26d14d7ca1bbe0d/pkg/sql/catalog/descs/uncommitted_descriptors.go#L201)):*
> ```Go
> 	}
> 	rawBytesInStorageAfterPendingWrite := val.TagAndDataBytes()
> 	uncommittedBuilder := mut.NewBuilder()
> ```

another option instead of the setter is a `NewBuilderWithRawBytes` method. I think I prefer that if you're open to it. 


<!-- Sent from Reviewable.io -->

Xiang-Gu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88844)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[`pkg/sql/catalog/dbdesc/database_desc_builder.go` line 151 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/88844#-NDUB5Ib06k7j7MHtydl:-NDUCDOC-l4Kea2-yG0v:b3vk5qy) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa3c59ccb2836ebd04ae89a4a26d14d7ca1bbe0d/pkg/sql/catalog/dbdesc/database_desc_builder.go#L151)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: you can do `make([]byte, 0, len(rawBytes))` to allocate the right size array underneath here and elsewhere.
</blockquote></details>

the reason I didn't do this is if `rawBytes` is nil, then we will end up with a 0-length slice, rather than also a `nil`.
___
*[`pkg/sql/catalog/descs/uncommitted_descriptors.go` line 201 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/88844#-NDUBN9bByIQGHzyaw_W:-NDUCV9e7HKcJWZ53xsC:bczh8om) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa3c59ccb2836ebd04ae89a4a26d14d7ca1bbe0d/pkg/sql/catalog/descs/uncommitted_descriptors.go#L201)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

another option instead of the setter is a `NewBuilderWithRawBytes` method. I think I prefer that if you're open to it. 
</blockquote></details>

If we want to *completely* get rid of this setter, we will need to change two places:
1. building from a protobuf: the current method is NewBuilderWithMVCCTimestamp. To replace it, we will need to plumb in the raw bytes there to construct a builder from (descpb.descriptor, mvccTimestamp, rawBytes);
2. building from a catalog.Descriptor (which is what we need here specifically): as you suggested, we can create a `NewBuilderWithRawBytes(rawBytes)` to replace the old `NewBuilder()` with a default value `nil`.

I'm open to it, in a different PR, if you prefer this.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88844)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @postamar, and @Xiang-Gu)

___
*[`pkg/sql/catalog/descs/uncommitted_descriptors.go` line 201 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/88844#-NDUBN9bByIQGHzyaw_W:-NDUQi7F69aAPt5oW4V8:b-gxjyrv) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa3c59ccb2836ebd04ae89a4a26d14d7ca1bbe0d/pkg/sql/catalog/descs/uncommitted_descriptors.go#L201)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

If we want to *completely* get rid of this setter, we will need to change two places:
1. building from a protobuf: the current method is NewBuilderWithMVCCTimestamp. To replace it, we will need to plumb in the raw bytes there to construct a builder from (descpb.descriptor, mvccTimestamp, rawBytes);
2. building from a catalog.Descriptor (which is what we need here specifically): as you suggested, we can create a `NewBuilderWithRawBytes(rawBytes)` to replace the old `NewBuilder()` with a default value `nil`.

I'm open to it, in a different PR, if you prefer this.
</blockquote></details>

Certainly no need to do it in this PR. I don't hate the setter, it's just that it always seems paired with construction. 


<!-- Sent from Reviewable.io -->
